### ISSUE TITLE
Release Tag

### ISSUE THREAD

**Issue by r3sus** [garyttierney/ghidra-batteries-included#60]
Hey
Readme can be updated to link 9.2.1 
https://ghidra-sre.org/releaseNotes_9.2.1.html
Are there significant changes to database format in dev/debugger branch? 
or it is possible to decrease Release Tag to 9.2.1 and keep it with latest public release? 
Would be useful in case I want to open db in public release, or even stay with it. 
Because for now I have db made in 9.3 dev version. Seems it will be a long way until it released publicly. 
Assume, it's possible to XML export database, but it's unwanted complication. 
Thanks

**Comment by garyttierney** [comment]
It doesn't look like the shifted pointers patch is going to be included any time soon (or at all) in Ghidra, so that is being dropped and I will look at a way to downgrade existing databases to before the patch was included.

I created a tag called 9.2.1 with binaries attached for archival purposes, it is a build of 9.2.1 with a few extra patches and the regular extension additions (still including shifted pointers, too).

**Comment by garyttierney** [comment]
After a bit of thought, I'm wondering if shifted pointers should be dropped now or we wait until Ghidra proposes an alternative. cc @astrelsky, do you have any thoughts on the above?

**Comment by astrelsky** [comment]
I pulled this from my local repo. This should be what you need. The upgrade prompt will actually be to "downgrade". This assumes there is no shift and just truncates the data.

```diff
commit 9feadfdb18dd300d364f910645e0e366b0aec1b2
Author: astrelsky <ajs222@njit.edu>
Date:   Sat Jan 2 09:45:03 2021 -0500

    Pointer fix

diff --git a/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/program/database/data/PointerDBAdapter.java b/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/program/database/data/PointerDBAdapter.java
index 255256c52..d332b3b39 100644
--- a/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/program/database/data/PointerDBAdapter.java
+++ b/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/program/database/data/PointerDBAdapter.java
@@ -63,7 +63,12 @@ abstract class PointerDBAdapter {
 			return new PointerDBAdapterV1(handle);
 		}
 		catch (VersionException e) {
-			return new PointerDBAdapterV0(handle);
+			try {
+				return new PointerDBAdapterV2(handle);
+			}
+			catch (VersionException e2) {
+				return new PointerDBAdapterV0(handle);
+			}
 		}
 	}
 
diff --git a/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/program/database/data/PointerDBAdapterV2.java b/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/program/database/data/PointerDBAdapterV2.java
index ded3e665d..c86562c53 100644
--- a/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/program/database/data/PointerDBAdapterV2.java
+++ b/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/program/database/data/PointerDBAdapterV2.java
@@ -15,13 +15,19 @@
  */
 package ghidra.program.database.data;
 
+import ghidra.util.exception.VersionException;
+
 import java.io.IOException;
 
 import db.*;
-import ghidra.util.exception.VersionException;
 
 class PointerDBAdapterV2 extends PointerDBAdapter {
+	final static int DOWNGRADE_VERSION = 3;
 	final static int VERSION = 2;
+	
+	static final Schema OLD_SCHEMA = new Schema(DOWNGRADE_VERSION, "Pointer ID", new Class[] {
+		LongField.class, LongField.class, ByteField.class, IntField.class },
+		new String[] { "Data Type ID", "Category ID", "Length", "Shift Offset" });
 
 	private Table table;
 
@@ -37,7 +43,7 @@ class PointerDBAdapterV2 extends PointerDBAdapter {
 			}
 			else if (table.getSchema().getVersion() != VERSION) {
 				int version = table.getSchema().getVersion();
-				if (version < VERSION) {
+				if (version < VERSION || table.getSchema().equals(OLD_SCHEMA)) {
 					throw new VersionException(true);
 				}
 				throw new VersionException(VersionException.NEWER_VERSION, false);
@@ -45,6 +51,21 @@ class PointerDBAdapterV2 extends PointerDBAdapter {
 		}
 	}
 
+	PointerDBAdapterV2(DBHandle handle) throws VersionException {
+
+		table = handle.getTable(POINTER_TABLE_NAME);
+		if (table == null) {
+			throw new VersionException("Missing Table: " + POINTER_TABLE_NAME);
+		}
+		else if (table.getSchema().getVersion() != VERSION && !table.getSchema().equals(OLD_SCHEMA)) {
+			int version = table.getSchema().getVersion();
+			if (version < VERSION) {
+				throw new VersionException(true);
+			}
+			throw new VersionException(VersionException.NEWER_VERSION, false);
+		}
+	}
+
 	@Override
 	DBRecord createRecord(long dataTypeID, long categoryID, int length) throws IOException {
 		long tableKey = table.getKey();
```

**Comment by garyttierney** [comment]
Ah, perfect, I'll apply this to the in-progress debugger branch with a note that support for downgrading will eventually be removed whenever real changes to the pointer schema occur.

**Comment by garyttierney** [comment]
9.2.1 tagged and shifted pointers downgrade is included in the latest 9.3 builds. Closing this now.
